### DISCUSSION TITLE
[A-L] (10 giugno 2019) Esercizio 5

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#90] [Esami A-L]
Ricorda che una passeggiata in un grafo diretto è un cammino che potrebbe ripetere i vertice e archi.
_`(lo ha scritto per se stesso perchè sennò si scordava AHAHAHAH)`_
Dare lo pseudocodice di un algoritmo che prenda in input un grafo diretto G, due vertici u,v, e un intero k e restituisca il numero di passeggiate distinte di lunghezza al massimo k da u a v.
L'algoritmo dovrebbe avere complessità $O(n\cdot m\cdot k)$

**Hint:** Si tratta di programmazione dinamica 😢

**Comment by CasuFrost** [comment]
![image](https://github.com/user-attachments/assets/eff023ac-9c00-4961-959b-761ad571c968)

**Comment by cornflexxx** [comment]
Definiamo la matrice $T_{n\times k}$ dove $T[i,q]\rightarrow$ numero di passeggiate da $u$ ad $i$ con $q$ archi.

$$T[i,q] := \left\\{ \begin{array}{cl}
1 & i = q = 0 \\
0 & q=0 \text{ oppure }\nexists u\to i \\
\displaystyle\sum_{p \in N_e(i)}T[p,q-1]& altrimenti \\
\end{array} \right.$$

```
Passeggiata_DP(G,u,v,k):
	T := matrice n,k [[0,...0],....]
	T[0,0] = 1
	for q = 0,...,n:
	  for i = 1,....,k:
	   for p in i.entranti:
	     T[i,q] += T[p,q-1]
	return ∑T[v,i] for i = 1 to k
	
```

**Reply by Elia-Belli** [reply]
Andrebbe eseguito prima il ciclo su $q$ , con il codice attuale stai cercando di calcolare le passeggiate di un singolo nodo con $1,2,\dots,k$ archi sequenzialmente, senza le informazioni sugli altri nodi.

**Comment by Elia-Belli** [comment]
## Programmazione Dinamica
$$T[e,w] = \text{numero di passeggiate $u\to w$ con esattamente e archi}$$
- $T[0,w] = 1$ se $w=u$
- $T[0,w] = 0$ se $w\neq u$

Esiste una passeggiata $u\to w$ di $e$ archi se è possibile estendere una passeggiata $u\to x$ di $e-1$ archi tramite $(x,u)$, e quindi si ottengono altre $T[e-1,x]$ possibili passeggiate per $w$, ciò per ogni nodo $x$ che ha appunto archi entranti in $w$.
- $T[w,e] = \sum \\{T[e-1,x]\ |\ \exists\ (x,w)\in E(G)\\}$

```python
def passeggiate(G,u,v,k):
    T = [[0]*n] * (k+1)      #tabella (k+1) x (n)
    T[0,0] = 1

    for e in range(k+1):   #O(k)
        for w in G.V:      #O(n)
        	for (x,w) in G.E:    #O(m)
        		T[e,w] += T[e-1,x]
    result = 0
    for e in range(k+1): result+= T[e,v]      #O(k)
    return result	
```

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/9fae0ef5-edf2-455d-be35-89b31724b28e)

**Comment by cek753** [comment]
<img width="1234" height="919" alt="image" src="https://github.com/user-attachments/assets/c0cec0b4-585b-4dfb-bb88-34474aa13d9c" />
